Change LLVM search order in cmake to be the default order.

arthurp opened the pull request:
Honor the build environment search order for LLVM versions. This may result in using a slightly older version of
llvm-support in cases where multiple are installed, however it enables the environment to select an older versions
intentionally if required. Conda does this and not allowing it to do so causes link errors.

witchel commented on libtsuba/src/RDG.cpp:
```diff
@@ -1725,39 +1754,52 @@ tsuba::Register(const std::string& name) {
+galois::Result<uint64_t>
+tsuba::internal::ParseVersion(const std::string& file) {
+  return ::ParseVersion(file);
+}
```
Does this code make me bad?  I thought it was a clever way to export an anonymous namespace function, but I don't really understand C++ namespaces.

ddn0 replied:
Stylistically, I'd just rename the function in the unnamed namespace to something else so you don't need to add the `::` global namespace prefix as a form of defensive programming, or in this case, since we are already exporting the function under the name `tsuba::internal::ParseVersion`, just redirecting all calls to that name, because we don't gain much from symbol hiding.

witchel replied:
Done.

witchel commented on libtsuba/include/tsuba/RDG.h:
```diff
@@ -106,23 +123,29 @@ class GALOIS_EXPORT RDGMeta {
+  std::shared_ptr<arrow::ChunkedArray> local_to_global_vector_{};
+  std::shared_ptr<arrow::ChunkedArray> global_to_local_keys_{};
+  std::shared_ptr<arrow::ChunkedArray> global_to_local_values_{};
 };
```
I'm a little unclear on the value of {} initialization for STL types.  I thought pointer types should have it, but std::vector doesn't need it. But I'm not sure.

ddn0 replied:
Yes, `T* member_` probably should be written as `T* member_{}` for safety but there is no need to have `T{}` (value initialization) for class types (as opposed to non-class types like T*, int, etc.). This includes `shared_ptr`. It acts like a pointer but is a class. So, you can leave the shared_ptr declarations without `{}`. 

Some more context:

https://github.com/KatanaGraph/katana/pull/431#discussion_r478556400

witchel replied:
Thank you.

tylershunt commented:
> So the command line is this piece of ambient state that gets passed around to Write() or Commit(), which means an unfortunate amount of threading it through DistGraph, Distribution, PropertyFileGraph, and of course RDG. 

For added Bleh, I learned (or I guess forced myself to remember) `PropertyFileGraph`'s interface is also depended on by the galois python package so there will likely be a build failure there (I think I fixed the linking problem getting it to work for my PR).

witchel commented:
> > So the command line is this piece of ambient state that gets passed around to Write() or Commit(), which means an unfortunate amount of threading it through DistGraph, Distribution, PropertyFileGraph, and of course RDG.
> 
> For added Bleh, I learned (or I guess forced myself to remember) `PropertyFileGraph`'s interface is also depended on by the galois python package so there will likely be a build failure there (I think I fixed the linking problem getting it to work for my PR).

I guess Bleh is in the eye of the beholder.  In some sense, the code is letting us know where it needs lineage information.  I'll talk to Tyler about how I can fix and test python.

ddn0 commented on libtsuba/tests/tsuba_gc.cpp:
```diff
@@ -94,140 +121,154 @@ FindVersions(const std::string& src_uri, uint32_t remaining_versions) {
+
+// Set difference
+std::unordered_set<std::string>
+SetDiff(std::unordered_set<std::string> a, std::unordered_set<std::string> b) {
```
Depending on your love of the STL:
https://en.cppreference.com/w/cpp/algorithm/set_difference

witchel replied:
That was my first thought, but my reading of the documentation was that the sets have to be sorted (e.g., Copies the elements from the sorted range [first1, last1) which are not found in the sorted range [first2, last2) to the range beginning at d_first.).

witchel replied:
Ok, ok, I rewrote it to use a set.

tylershunt reviewed (approved):
Overall, I'm very into this change!

tylershunt commented:
> > > So the command line is this piece of ambient state that gets passed around to Write() or Commit(), which means an unfortunate amount of threading it through DistGraph, Distribution, PropertyFileGraph, and of course RDG.
> > 
> > 
> > For added Bleh, I learned (or I guess forced myself to remember) `PropertyFileGraph`'s interface is also depended on by the galois python package so there will likely be a build failure there (I think I fixed the linking problem getting it to work for my PR).
> 
> I guess Bleh is in the eye of the beholder. In some sense, the code is letting us know where it needs lineage information. I'll talk to Tyler about how I can fix and test python.

Oh yeah I meant `bleh` as in "yay more stimulating threading work" not `bleh` as in "I don't think we should do this"

witchel commented:
> > > > So the command line is this piece of ambient state that gets passed around to Write() or Commit(), which means an unfortunate amount of threading it through DistGraph, Distribution, PropertyFileGraph, and of course RDG.
> > > 
> > > 
> > > For added Bleh, I learned (or I guess forced myself to remember) `PropertyFileGraph`'s interface is also depended on by the galois python package so there will likely be a build failure there (I think I fixed the linking problem getting it to work for my PR).
> > 
> > 
> > I guess Bleh is in the eye of the beholder. In some sense, the code is letting us know where it needs lineage information. I'll talk to Tyler about how I can fix and test python.
> 
> Oh yeah I meant `bleh` as in "yay more stimulating threading work" not `bleh` as in "I don't think we should do this"

Graduate school is the process by which `bleh `comes to be known as "yay, more stimulating threading work."

witchel commented on libgalois/include/galois/graphs/PropertyFileGraph.h:
```diff
@@ -125,28 +127,41 @@ class GALOIS_EXPORT PropertyFileGraph {
+    } else {
+      GALOIS_LOG_DEBUG(
+          "set_partition_metata initialized! state {} num_nodes: {}",
+          rdg_.part_metadata_.state, rdg_.part_metadata_.num_nodes_);
```
Yah, will do.

witchel commented on libgalois/test/property-file-graph.cpp:
```diff
@@ -46,7 +46,7 @@ TestRoundTrip() {
   std::string rdg_file{temp_dir};
 
-  auto write_result = g->Write(rdg_file);
+  auto write_result = g->Write(rdg_file, "property-file-graph");
```
No problemo.  The program is simple enough I just updated it to compute the command line and pass it along.

witchel commented on libtsuba/include/tsuba/RDG.h:
```diff
@@ -44,6 +44,20 @@ class GALOIS_EXPORT RDGFile {
+class RDGLineage {
+public:
+  std::string command_line_{};
+  RDGLineage() = default;
```
Cool article.

witchel commented on libtsuba/src/RDG.cpp:
```diff
@@ -653,7 +680,11 @@ DoReadMetadata(const std::string& partition_path) {
+    if (!res) {
+      res = ReadMetadataJson(partition_path);
+    }
+    return res;
```
Good catch.

arthurp committed:
Change LLVM search order in cmake to be the default order.

Honor the build environment search order for LLVM versions. This may result in using a slightly older version of
llvm-support in cases where multiple are installed, however it enables the environment to select an older versions
intentionally if required. Conda does this and not allowing it to do so causes link errors.

GioLomia commented:
Looks good. The reversed logic makes much more sense IMHO. I'm wondering why it wasn't like that in the first place.

GioLomia reviewed (approved):
LGDTM

ddn0 commented on cmake/Modules/BuildCommon.cmake:
```diff
@@ -285,20 +285,18 @@ if (NOT MySQL_FOUND)
-  endif ()
-endforeach ()
-if (LLVM_FOUND)
+# Honor the build environment search order for LLVM versions. This may result in using a slightly older version of
```
~"It [i.e., the search order] may result"~ edit: brain fart: ignore this comment
